[EDR Workflows] Initialize argument selector value

tomsonpl opened the pull request:
## Description

Previously, #204965 introduced a `string` based Selector Component, however the Response Console's framework wasn't adjusted to fully support it:

- **Selector Components Not Restoring State from History:**  
  When users selected a command from history (e.g., custom script execution), the selector UI (such as script dropdowns) could not restore its value from the previously entered command text. This left the UI out of sync with the command, confusing users and breaking workflows.

- **Broken Paste Functionality:**  
  Pasting a full command (including arguments meant for selectors) into the console did not populate the selector UI components. As a result, users who pasted commands from documentation or shared snippets would see empty or incorrect selector states, making it difficult to execute or edit those commands.

These issues stemmed from the lack of a unified mechanism for synchronizing between the command text and selector state. Command-specific logic was difficult to maintain and extend.

---

## How was it resolved?

** Added `argState` Infrastructure:**
- Enhanced `Command` interface with `argState` property to track selector component state
- Modified command execution pipeline to preserve selector state across history and paste operations

** Smart Paste Processing:**
- Implemented `detectAndPreProcessPastedCommand()` function that extracts values from pasted commands like `runscript --ScriptName="test.ps1"`
- Separates embedded values into selector state while cleaning the command text for display

** History State Restoration:**
- Updated `InputHistoryItem` to include `argState` property
- Modified history selection to restore both command text and selector UI state

---

## Additional Architectural Improvements

Based on code review feedback, this PR also addresses architectural concerns with the `CustomScriptSelector` component:

- **🔧 Standard React Component:**  
  Converted from function-returning-component pattern (`CustomScriptSelector('agentType')`) to standard React component (`<CustomScriptSelector />`)

- **🔧 Console Framework Encapsulation:**  
  Removed direct `useConsoleStateDispatch()` coupling and replaced with proper callback-based architecture (`onChange`, `requestFocus`)

- **🔧 Command Metadata Integration:**  
  Agent type now extracted from `command.commandDefinition.meta?.agentType` instead of function parameters, following console framework patterns

- **🔧 Enhanced Interface:**  
  Added `command: Command` and `requestFocus?: () => void` props to `CommandArgumentValueSelectorProps` for better component integration

These changes ensure the console remains a generic, maintainable framework while providing clean interfaces for selector components.

---

## Result

- Selector UI components now always reflect the correct state when loading from history or after pasting a command (for commands where this is supported).
- Users can confidently copy, paste, and re-run complex commands with selector arguments.
- The codebase is more maintainable and ready for future extension.
- Console framework maintains proper encapsulation with well-defined component interfaces.

https://github.com/user-attachments/assets/bf7a52d4-c8de-4363-8b53-af248bea8366


Closes: https://github.com/elastic/kibana/issues/225269
Closes: https://github.com/elastic/kibana/issues/225407

<!--ONMERGE {"backportTargets":["8.19","9.1"]} ONMERGE-->

paul-tavares committed:
Store argState with history items and pass them to argument value selectors

tomsonpl committed:
fix

tomsonpl committed:
fix pasting

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

tomsonpl committed:
remove case specific logic from the console component.

tomsonpl committed:
specify logic to handle file upload fresh start

tomsonpl commented:
/ci

tomsonpl committed:
fix type

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

tomsonpl committed:
tests and types

tomsonpl commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint_all_files --no-cache --fix'

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318204)
* Commit: ef3bae16e269d8057f3300eafb1337ac7dbde5a8

### Failed CI Steps
* [Linting](https://buildkite.com/elastic/kibana-pull-request/builds/318204#0197f983-84db-4dc3-8306-b28ecf406e58)

### History
* :broken_heart: [Build #318040](https://buildkite.com/elastic/kibana-pull-request/builds/318040) failed d4896caf188b359839fb152c141d42a2cee7de4e

cc @tomsonpl

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318204","number":318204,"commit":"ef3bae16e269d8057f3300eafb1337ac7dbde5a8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318040","number":318040,"commit":"d4896caf188b359839fb152c141d42a2cee7de4e"}],"number":318204}
buildkite-pr-comment-->

tomsonpl commented:
/ci

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/318229)
* Commit: 695d449fd80dcee4391c081b42236228f6df8393
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-227463)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/5549d70cc9105824f86b323f222f8e52bb8b08e7) | [after](https://github.com/elastic/kibana/commit/695d449fd80dcee4391c081b42236228f6df8393) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.8MB | 9.8MB | +826.0B |



### History
* :broken_heart: [Build #318204](https://buildkite.com/elastic/kibana-pull-request/builds/318204) failed ef3bae16e269d8057f3300eafb1337ac7dbde5a8
* :broken_heart: [Build #318040](https://buildkite.com/elastic/kibana-pull-request/builds/318040) failed d4896caf188b359839fb152c141d42a2cee7de4e

cc @tomsonpl

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318229","number":318229,"commit":"695d449fd80dcee4391c081b42236228f6df8393"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318204","number":318204,"commit":"ef3bae16e269d8057f3300eafb1337ac7dbde5a8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/318040","number":318040,"commit":"d4896caf188b359839fb152c141d42a2cee7de4e"}],"number":318229}
buildkite-pr-comment-->

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

paul-tavares commented on x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/console_state/state_update_handlers/handle_input_area_state.ts:
```diff
@@ -103,15 +119,38 @@ export const handleInputAreaState: ConsoleStoreReducer<InputAreaStateAction> = (
+        const fullCommand = newTextEntered + newRightOfCursor;
+
+        // Pre-process pasted commands to handle selector arguments like --ScriptName
+        const preProcessResult = detectAndPreProcessPastedCommand(fullCommand, state.commands);
```
This should only be done if the user pasted the value in the console area. By the time that data reaches the state handling logic here, it should not be concerned with "pasted" values. I should just `update` the input area state.

I think the best place to handle this is here: `x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/command_input/command_input.tsx:152`

This is the place where all user keyboard interactions while focused on the input area are handled.

Note that you may need to ensure the `paste` event is being reported correctly by the `input_capture` component now that we are actually needing to process on that -  here: `x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/command_input/components/input_capture.tsx:182`

paul-tavares commented on x-pack/solutions/security/plugins/security_solution/public/management/components/console/service/parsed_command_input.ts:
```diff
@@ -25,6 +28,9 @@ const parseInputString = (rawInput: string): ParsedCommandInput => {
       : input.substring(inputFirstSpacePosition).trim().split(/--/);
 
+  // Find command definition for selector argument handling
+  const commandDef = commandDefinitions.find((def) => def.name === response.name);
```
Same for changes in this module. its purpose is to only focus on parsing the text the user entered (this is the input parser I often refer to)... It should not be responsible for associating that text entered/parsed to a known command/argument.

I think that most of these changes should also be moved to `x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/command_input/command_input.tsx:152` 

Keep the separation of concerns will make it easier to switch out this parser with something else in the future.

paul-tavares commented on x-pack/solutions/security/plugins/security_solution/public/management/components/console/types.ts:
```diff
@@ -67,6 +67,11 @@ export interface CommandArgDefinition {
+  /**
+   * If defined, the selector will use `true` as the default value which results in empty array as args value
+   */
+  selectorStringDefaultValue?: boolean;
```
I got a bit confused by this new prop's name. Could you describe what its purpose is?

Is it used only when processing commands from Paste action or selection from input history? and only during `SelectorComponent` initialization? or checked everytime the user input changes (ex. on typing).

paul-tavares reviewed (commented):
Sorry for the delay in reviewing your latest changes. It takes me a while to review this one and I was not able to find time earlier this week. 

Thanks for all the work you are doing on this. I left some comments and also did manual testing locally. `upload` seems to work as expected, but encountered some issue with MS `runscript`

- pasting something like `runscript --ScriptName="invalid.sh"` and pressing `[ENTER]` seems to submit the response action via API. I would have expected the value to `--ScriptName` to be validated (ensure it is a valid script name)
- When pasting the command with a valid script name (ex. `runscript --ScriptName="hello.sh"`), that script name is not showign up as "selected" if I then click the Script Selector.



**Maybe an issue/limitation:**

I also noticed that it seem as if you can not execute any other `runscript` (or, I assume, no other `runliveresponse` MS command) while the prior one is pending. Were you aware of this? (appologies if perhaps we covered this during initial grooming for `runscript` - I don't recall).  If that is the case, we should discuss at one of our future sync meetings with Raquel and others. If my assumption is true, I wonder if we should be handing this condition more gracefully for the users in the console.

Here is the error I got:

```
Attempt to send [runScript] to Microsoft Defender for Endpoint failed: Status code: 400. Message: API Error: [Bad Request] Request failed with status code 400
URL called:[post] https://api.securitycenter.windows.com/api/machines/75d7cc8ff4520e8c583f219e516754849c4ec6c7/runliveresponse
Response body: {"error":{"code":"ActiveRequestAlreadyExists","message":"Action is already in progress, active action id: 6a3cb5d3-a567-453f-8f26-4b2a94873188","target":"|00-d8daa111f7ba6726bb1da04bf3ff9205-be3471cbb817b716-01.9d9452fb_1.1."}}
```



**Possible new issue**:

Also - one other thing I have forgotten to tell you wiht prior reviews: I have noticed this error is now showing up on the kibana logs:

```
[2025-07-19T12:43:00.097-04:00][ERROR][plugins.securitySolution.MicrosoftDefenderEndpointAgentStatusClient] Attempt to calculate isolate state for Microsoft Defender hosts generated the following errors:
Cannot read properties of undefined (reading 'type')
```

Are you already aware of this issue?

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

# Conflicts:
#	x-pack/solutions/security/plugins/security_solution/public/management/components/console_argument_selectors/custom_scripts_selector/custom_script_selector.test.tsx
#	x-pack/solutions/security/plugins/security_solution/public/management/components/console_argument_selectors/custom_scripts_selector/custom_script_selector.tsx

tomsonpl committed:
resolve conficts

tomsonpl committed:
move logic to command_input

tomsonpl committed:
rename flag name

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

tomsonpl commented:
@ashokaditya Hey! Since you're back from PTO  - I removed @szwarckonrad from reviewers and added you, would you find some time to take a look at this one? :) 
Thanks!

tomsonpl committed:
fix cannot read type of undefined error

tomsonpl committed:
Merge remote-tracking branch 'origin/initialize-arg-selector' into initialize-arg-selector

paul-tavares commented on x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/console_state/state_update_handlers/handle_input_area_state.ts:
```diff
@@ -77,6 +78,18 @@ export const handleInputAreaState: ConsoleStoreReducer<InputAreaStateAction> = (
               display: payload.display ?? payload.command,
+              // We only store the `value` and `valueText`. `store` property of each argument's state
+              // is component instance specific data.
+              argState: Object.entries(payload.argState || {}).reduce(
```
Just FYI (for the future, case its needed). The console history also supports migration if we ever need it. I have used it once in the past (version is currently at `2`), but for this change here I don't think we need it since the data is optional.

https://github.com/elastic/kibana/blob/da25d13a2ac5be97216e202bf5d6bda6ff920b3b/x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/console_state/hooks/use_stored_input_history.ts#L79-L93

tomsonpl commented on x-pack/solutions/security/plugins/security_solution/public/management/components/console/components/command_input/command_input.tsx:
```diff
@@ -76,6 +79,94 @@ const CommandInputContainer = styled.div`
+ * - cleanedCommand: "runscript --ScriptName"
+ * - extractedArgState: { ScriptName: [{ value: "test.ps1", valueText: "test.ps1" }] }
+ */
+const detectAndPreProcessPastedCommand = (
```
Good catch, sorry a leftover from refactoring. Thanks!

tomsonpl committed:
remove duplicated code and move to utils

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/324437)
* Commit: 76741f485c27d0dc0be2e46da00ef87d6b0e202c
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-227463)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/7a129328fb7bd1f7732216517ca011d7266e64bb) | [after](https://github.com/elastic/kibana/commit/76741f485c27d0dc0be2e46da00ef87d6b0e202c) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 7694 | 7695 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/7a129328fb7bd1f7732216517ca011d7266e64bb) | [after](https://github.com/elastic/kibana/commit/76741f485c27d0dc0be2e46da00ef87d6b0e202c) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.8MB | 9.8MB | +853.0B |



### History
* :green_heart: [Build #324292](https://buildkite.com/elastic/kibana-pull-request/builds/324292) succeeded 326c2dacf3d627922fa816c28bea3fd5a66ec383
* :green_heart: [Build #323600](https://buildkite.com/elastic/kibana-pull-request/builds/323600) succeeded c12156a83ab014a3da05d0fe3f17aa21bc867f3d
* :broken_heart: [Build #323404](https://buildkite.com/elastic/kibana-pull-request/builds/323404) failed f6e6a55e83aa813b638704a7b25d93b507323178
* :broken_heart: [Build #323184](https://buildkite.com/elastic/kibana-pull-request/builds/323184) failed 4d7f625e6d2c00e985bfe4a56691b6500744e343
* :broken_heart: [Build #322899](https://buildkite.com/elastic/kibana-pull-request/builds/322899) failed f74bfaf8a5ce4e9509641729101c922daa5f28fe
* :green_heart: [Build #318739](https://buildkite.com/elastic/kibana-pull-request/builds/318739) succeeded aa4de077acdfe9079f9afa03c49563615590564e

cc @tomsonpl

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324437","number":324437,"commit":"76741f485c27d0dc0be2e46da00ef87d6b0e202c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324292","number":324292,"commit":"326c2dacf3d627922fa816c28bea3fd5a66ec383"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323600","number":323600,"commit":"c12156a83ab014a3da05d0fe3f17aa21bc867f3d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323404","number":323404,"commit":"f6e6a55e83aa813b638704a7b25d93b507323178"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323184","number":323184,"commit":"4d7f625e6d2c00e985bfe4a56691b6500744e343"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/322899","number":322899,"commit":"f74bfaf8a5ce4e9509641729101c922daa5f28fe"}],"number":324437}
buildkite-pr-comment-->

tomsonpl committed:
Merge branch 'main' into initialize-arg-selector

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/325414)
* Commit: e39d9891ad4fa97b033a71fcc3242669509147eb
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-227463)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/ea7a63230b61550659435dc7d221f5b203d654e3) | [after](https://github.com/elastic/kibana/commit/e39d9891ad4fa97b033a71fcc3242669509147eb) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 7611 | 7612 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/ea7a63230b61550659435dc7d221f5b203d654e3) | [after](https://github.com/elastic/kibana/commit/e39d9891ad4fa97b033a71fcc3242669509147eb) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 9.8MB | 9.8MB | +853.0B |



### History
* :green_heart: [Build #324437](https://buildkite.com/elastic/kibana-pull-request/builds/324437) succeeded 76741f485c27d0dc0be2e46da00ef87d6b0e202c
* :green_heart: [Build #324292](https://buildkite.com/elastic/kibana-pull-request/builds/324292) succeeded 326c2dacf3d627922fa816c28bea3fd5a66ec383
* :green_heart: [Build #323600](https://buildkite.com/elastic/kibana-pull-request/builds/323600) succeeded c12156a83ab014a3da05d0fe3f17aa21bc867f3d
* :broken_heart: [Build #323404](https://buildkite.com/elastic/kibana-pull-request/builds/323404) failed f6e6a55e83aa813b638704a7b25d93b507323178
* :broken_heart: [Build #323184](https://buildkite.com/elastic/kibana-pull-request/builds/323184) failed 4d7f625e6d2c00e985bfe4a56691b6500744e343
* :broken_heart: [Build #322899](https://buildkite.com/elastic/kibana-pull-request/builds/322899) failed f74bfaf8a5ce4e9509641729101c922daa5f28fe

cc @tomsonpl

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/325414","number":325414,"commit":"e39d9891ad4fa97b033a71fcc3242669509147eb"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324437","number":324437,"commit":"76741f485c27d0dc0be2e46da00ef87d6b0e202c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/324292","number":324292,"commit":"326c2dacf3d627922fa816c28bea3fd5a66ec383"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323600","number":323600,"commit":"c12156a83ab014a3da05d0fe3f17aa21bc867f3d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323404","number":323404,"commit":"f6e6a55e83aa813b638704a7b25d93b507323178"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/323184","number":323184,"commit":"4d7f625e6d2c00e985bfe4a56691b6500744e343"}],"number":325414}
buildkite-pr-comment-->
